### PR TITLE
Add More Picotrav Variable Ordering Heuristics

### LINKED ISSUE #21: Add 'ZIP' variable ordering to Picotrav
Another simple variable ordering that would be of interest is a _zipped_ ordering. This is specifically of interest for the circuits with two inputs, i.e. most of the _arithmetic_ circuits of _EPFL_. Assuming the input is given as `a[0]`, `a[1]`, ..., `a[n-1]`, `b[0]`, `b[1]`, ..., `b[n-1]` we would like to derive the ordering `a[0]`, `b[0]`, `a[1]`, `b[1]`, ..., `a[n-1]`, `b[n-1]`.

This can be done with a simple for-loop with pointers starting at the first and middle-most element.

### PR DESCRIPTION
**Pull request by nhusung**
This adds the fanin variable ordering heuristic as well as the ZIP heuristic (#21). To respect both circuits in the variable ordering, I had to do a major refactoring. Both nets now contain a mapping from node names (strings) to IDs (integers), which are just indices into a shared node store (a `std::vector`). Also, we match the inputs and outputs if two circuits are given based on their names. I’m not sure, but I think we previously relied on the order in which the inputs/outputs are given. I suppose name-based matching is more desired in cases where it matters. But I could also introduce a command line option, if you want.

In my tests, the ZIP heuristic worked well for, e.g., the adder benchmark, but does not work very well in general. I’ve seen cases where the fanin heuristic significantly outperforms the level DFS heuristic. For instance, verifying that `mem_ctrl_depth_2023` and `mem_ctrl` of the EPFL benchmark suite agree takes only 13 seconds on my laptop (using CUDD and the fanin heuristic).

There are some minor fixes and additions to the CNF benchmark included in this pull request. Is this fine by you or should I split this into two pull requests?

Closes #21

### CODE REVIEW

**Comment by SSoelvsten** [comment]
> I’m not sure, but I think we previously relied on the order in which the inputs/outputs are given. I suppose name-based matching is more desired in cases where it matters. But I could also introduce a command line option, if you want.

Indeed, this was the case. I had examples from *EPFL* or another place where the names did not match. And so, the order is everything you can base it off of. Hence, I'm not sure whether such a change is "correct" with respect to the input format (except if there is something in the official documentation on *BLIF* that says its ok?). It seems odd, but one could expect a different input/output to be generated with the same name (even if it is another one).

**Comment by SSoelvsten** [comment]
> In my tests, the ZIP heuristic worked well for, e.g., the adder benchmark, but does not work very well in general. 

That is to be expected. Zipping should work well when input variables are bits of two numbers on which you do arithmetic operations, e.g. the *adder*. In other cases, such as `mem_ctrl` one should probably expect it doesn't work.

**Comment by SSoelvsten** [comment]
Good suggestion. From the outside you could indeed do it as an `-m name` or `-m order` which then internally translates into a `match_names` Boolean flag.

**Comment by SSoelvsten** [comment]
Purely out of interest. For the CNF benchmark, are you running into the *satcount* operation takes too long time?

**Comment by nhusung** [comment]
> Purely out of interest. For the CNF benchmark, are you running into the _satcount_ operation takes too long time (since you added [24c2f64](https://github.com/SSoelvsten/bdd-benchmark/commit/24c2f64c30a466b8f7a8aef71ea192a35b05f9fa))?

Sylvan and CUDD are my troublemakers. With CUDD, I’ve seen cases where a `std::logic_error` is thrown. I tried to use Sylvan for the same files (DD construction was fine, about 130 k nodes and 2 k variables), but SAT counting did not finish within a minute or so. I’m quite sure that I was able to reproduce these issues earlier today on my local machine, but just now, it seems to work. Maybe, there is some kind of issue with too large floating point numbers …

**Comment by nhusung** [comment]
The `-m` option is implemented and documented now. I’m not really satisfied by the help text; the description is a bit too long:

```
Picotrav Benchmark
-------------------------------------------------------------------------------
Usage:  -flag      [default] Description
-------------------------------------------------------------------------------
        -h                   Print this information

-------------------------------------------------------------------------------
BDD Package options:
        -M MiB       [128]    Amount of memory (MiB)
        -P THREADS   [1]      Worker thread count
        -R                    Enable dynamic variable reordering
        -T TEMP_PTH  [/tmp]   Filepath for temporary files on disk

-------------------------------------------------------------------------------
Benchmark options:
        -f PATH               Path to '.blif' file(s)
        -m order|name [order] Matching of circuit inputs and outputs
        -o ORDER     [input]  Variable order to derive from first circuit
```

**Comment by nhusung** [comment]
While testing the `-m` option, I also noticed that specifying an output that also is just an input leads to a SEGFAULT (or failed assertions in debug builds) in `verify_outputs()`. The problem is that inputs are not inserted into the cache. Is this “intended” or should I fix it?

**Comment by SSoelvsten** [comment]
> Sylvan and CUDD are my troublemakers. With CUDD, I’ve seen cases where a std::logic_error is thrown. I tried to use Sylvan for the same files (DD construction was fine, about 130 k nodes and 2 k variables), but SAT counting did not finish within a minute or so. I’m quite sure that I was able to reproduce these issues earlier today on my local machine, but just now, it seems to work. Maybe, there is some kind of issue with too large floating point numbers …
>
> Here’s a log from the benchmarking machine with CUDD ...

Interesting.

- Sylvan changes its performance quite significantly when you change the size of its computation cache. Maybe try to increase it or decrease it?
- That error of CUDD I cannot immediately answer where should arise. The CUDD source code nor its adapter seems to have a `std::logic_error` with such a `what()` message.

**Comment by SSoelvsten** [comment]
> ```
> Benchmark options:
>         -f PATH               Path to '.blif' file(s)
>         -m order|name [order] Matching of circuit inputs and outputs
>         -o ORDER     [input]  Variable order to derive from first circuit
> ```

To be consistent, `order|name` should be a singular word that is its generic category, e.g. `MATCH`.

**Comment by nhusung** [comment]
> It is intended, but not necessary. Creating the ith-variable takes O(1) time in any BDD package. Hence, the cache shouldn't be necessary (and the cache becomes slightly smaller and hence may perform better). But, the code may indeed become simpler if this is changed.

Sorry, I should have been more precise. With “intended” I was referring to the SEGFAULT, which already happened before my refactoring. The question was rather whether I should fix it. Now, I committed a fix which does not modify the cache. It’s not that concise, but also not overly cluttered, IMHO.

**Comment by nhusung** [comment]
Regarding the help strings (`-h`): As a user, I would find it convenient to be shown the options I could pick from. If I did not miss anything, we never present these options in the help text. On the other hand, describing the options in both the help text and the README increases the maintenance effort a little. (There would be to options: just naming the options or even describing them as in the README.) Just to be on the same page: Your current stance is that the convenience is not worth the extra effort? Or should I, for instance, describe `-m` as “Matching of circuit inputs and outputs (`order|name`)”?

**Comment by SSoelvsten** [comment]
> Sorry, I should have been more precise. With “intended” I was referring to the SEGFAULT, which already happened before my refactoring.

Ah, my bad. Yes that ought to be fixed! It just didn't occur to me one should want to propagate an input signal. But, now you point it out, if course that may happen! The fix in [b286300](https://github.com/SSoelvsten/bdd-benchmark/pull/136/commits/b28630092841c9c0aeed4a87d7015485551de986) looks good to me.

**Comment by SSoelvsten** [comment]
> As a user, I would find it convenient to be shown the options I could pick from.

I see the point. I just have had bad experiences with having to maintain two or three different sets of documentation. They very easily go out-of-sync. For example, the `make help` that existed until you advocated using CMake directly, was out of date when Tom van Dijk had to use it.

**Comment by SSoelvsten** [comment]
I'll get to do a proper code review soon, I promise. I have quite a few things to deal with this week.

File: src/picotrav.cpp [hunk 1]
```diff
@@ -872,13 +1001,22 @@ construct_net_bdd(const std::string& filename,
...
-verify_outputs(const net_t& net_0,
+verify_outputs(Adapter& adapter,
+               const net_t& net_0,
                const bdd_cache<Adapter>& cache_0,
                const net_t& net_1,
                const bdd_cache<Adapter>& cache_1)
 {
-  assert(net_0.outputs_in_order.size() == cache_0.size());
-  assert(net_1.outputs_in_order.size() == cache_1.size());
+  // The cache should exactly have exactly as many entries as there are outputs
```
Comment (SSoelvsten) [inline src/picotrav.cpp:1010]: Exactly one "exactly" too many.

File: src/picotrav.cpp [hunk 2]
```diff
@@ -963,6 +1092,42 @@ run_picotrav(int argc, char** argv)
...
+          if (net_1.nodes[id_1].is_output) {
+            net_1.outputs_in_order[i] = id_1;
+            continue; // success, continue with the next input
+          }
+        }
+
+        std::cerr << "Output '" << name << "' is present in '" << file_0 << "' but not in '"
+                  << file_1 << "'\n";
+        return -1;
+      }
```
Comment (SSoelvsten) [inline src/picotrav.cpp:1129]: This may be moved into a separate function - it is arguably "too low level" for the *main* function.

**Review by SSoelvsten — COMMENTED** [review]
Looks good. I found some minor things. But as such it is ready for a merge.

**Comment by nhusung** [comment]
Thank you for the review! The open points should be resolved now.

**Pull request merged**